### PR TITLE
Stabilize `use_extern_macros`

### PR DESCRIPTION
**Pull request by petrochenkov**
Closes https://github.com/rust-lang/rust/issues/35896

### CODE REVIEW

**Comment by kennytm** [comment]
cc @rust-lang/infra Crater run requested.

**Comment by pietroalbini** [comment]
Crater run (check only) started.

**Comment by pietroalbini** [comment]
Hi @kennytm (crater requester, PR reviewer)! Crater results are at: http://cargobomb-reports.s3.amazonaws.com/pr-50911/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by kennytm** [comment]
6 regressions.

* **verilog**: Seems to be spurious ("Read-only file system" cc @pietroalbini). 
* **npbot**: This PR seems to have broken its use of **structopt-^0.2.5**. Legit.
* ***The rest***: All due to `#[macro_use]` no longer needed, which causes `#[deny(warnings)]` to error out. Legit but expected.

<details><summary>Relevant source code in <b>npbot-1.0.0</b></summary>

```toml
[dependencies]
structopt = "0.2.5"
```

```rust
#[macro_use]
extern crate quicli;

// ...

use structopt::StructOpt;
use quicli::prelude::*;

// ...

#[derive(StructOpt)]
pub(crate) struct Opt {
    #[structopt(short = "d", long = "debug", help = "only use the local, println!() sink")]
    debug: bool,
    #[structopt(long = "verbose", short = "v", parse(from_occurrences))]
    verbose: u8,

    // for the SoundTouch source
    #[structopt(long = "soundtouch-host")]
    soundtouch_host: Option<String>,

    // for the last.fm source
    #[structopt(long = "lastfm-api-key")]
    lastfm_api_key: Option<String>,
    #[structopt(long = "lastfm-username")]
    lastfm_username: Option<String>,
}
```

Note the lack of `#[macro_use] extern crate structopt`.
</details>

<details><summary>Error log</summary>

```
May 22 22:49:57.928 INFO kablam! error[E0658]: The attribute `structopt` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
May 22 22:49:57.928 INFO kablam!   --> src/main.rs:41:5
May 22 22:49:57.928 INFO kablam!    |
May 22 22:49:57.928 INFO kablam! 41 |     #[structopt(short = "d", long = "debug", help = "only use the local, println!() sink")]
May 22 22:49:57.928 INFO kablam!    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 22 22:49:57.928 INFO kablam!    |
May 22 22:49:57.929 INFO kablam!    = help: add #![feature(custom_attribute)] to the crate attributes to enable
May 22 22:49:57.929 INFO kablam! 
May 22 22:49:57.929 INFO kablam! error[E0658]: The attribute `structopt` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
May 22 22:49:57.929 INFO kablam!   --> src/main.rs:43:5
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam! 43 |     #[structopt(long = "verbose", short = "v", parse(from_occurrences))]
May 22 22:49:57.929 INFO kablam!    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam!    = help: add #![feature(custom_attribute)] to the crate attributes to enable
May 22 22:49:57.929 INFO kablam! 
May 22 22:49:57.929 INFO kablam! error[E0658]: The attribute `structopt` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
May 22 22:49:57.929 INFO kablam!   --> src/main.rs:47:5
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam! 47 |     #[structopt(long = "soundtouch-host")]
May 22 22:49:57.929 INFO kablam!    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam!    = help: add #![feature(custom_attribute)] to the crate attributes to enable
May 22 22:49:57.929 INFO kablam! 
May 22 22:49:57.929 INFO kablam! error[E0658]: The attribute `structopt` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
May 22 22:49:57.929 INFO kablam!   --> src/main.rs:51:5
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam! 51 |     #[structopt(long = "lastfm-api-key")]
May 22 22:49:57.929 INFO kablam!    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 22 22:49:57.929 INFO kablam!    |
May 22 22:49:57.929 INFO kablam!    = help: add #![feature(custom_attribute)] to the crate attributes to enable
May 22 22:49:57.929 INFO kablam! 
May 22 22:49:57.930 INFO kablam! error[E0658]: The attribute `structopt` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
May 22 22:49:57.930 INFO kablam!   --> src/main.rs:53:5
May 22 22:49:57.930 INFO kablam!    |
May 22 22:49:57.930 INFO kablam! 53 |     #[structopt(long = "lastfm-username")]
May 22 22:49:57.930 INFO kablam!    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
May 22 22:49:57.930 INFO kablam!    |
May 22 22:49:57.930 INFO kablam!    = help: add #![feature(custom_attribute)] to the crate attributes to enable
May 22 22:49:57.930 INFO kablam! 
May 22 22:49:57.931 INFO kablam! error: aborting due to 5 previous errors
```

</details>

**Comment by petrochenkov** [comment]
I'll look what happens with `npbot`.

File: src/librustc_resolve/macros.rs
Comment (petrochenkov) [inline src/librustc_resolve/macros.rs]: FIXME: `possible_time_travel` is now unused

**Comment by petrochenkov** [comment]
It looks like `npbot` legitimately encounters a case of indeterminate resolution when macro expansion cannot progress and generates dummy fragments to recover.

So the error should be something like "cannot determine resolution for the derive macro `StructOpt`".

However, due to a scary hole in our macro resolution checking such dummy expansions can happen silently without generating any errors.
So we are seeing only secondary errors about unknown `structopt` attributes instead of the actual error about unresolved `StructOpt`.
I'll prepare a fix.

**Comment by petrochenkov** [comment]
Fixed in https://github.com/rust-lang/rust/pull/51145

**Comment by alexcrichton** [comment]
r=me, but I'll defer to...

r? @nikomatsakis 

pending https://github.com/rust-lang/rust/issues/35896#issuecomment-393231742

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@petrochenkov I'm a bit confused about what would be being stabilized by this PR.

I presume it is the full set of features described by @withoutboats in [this comment here](https://github.com/rust-lang/rust/issues/35896#issuecomment-378271859) — and in particular, it includes `macro_rules!` macros, right?

It seems like this *may* preclude various language changes to try and mitigate the `use` problem? We had intended to hold a @rust-lang/lang meeting to go over the implications in detail, and I'd love to try and puzzle them out (as it happens, I sort of learn towards accepting the feature as is, for basically the same reasons you described [here](https://github.com/rust-lang/rust/issues/35896#issuecomment-392393413) — but I'd like to understand the full implications first).

**Comment by petrochenkov** [comment]
@nikomatsakis 
Sorry for the delay.
Yes, this PR intends to stabilize the full set of features described by @withoutboats in https://github.com/rust-lang/rust/issues/35896#issuecomment-378271859.
(I'll answer to the recent discussion on macro helpers in https://github.com/rust-lang/rust/issues/35896 thread.)

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/51496

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/51496 was merges, so let's stabilize this!

ping @nikomatsakis

**Comment by joshtriplett** [comment]
I think we need a fix for https://github.com/rust-lang/rust/pull/51496#issuecomment-396326489 before stabilizing this.

I can live with not being able to `use my_module::my_macro;`; however, if a crate switches to using `$crate::` inside its own macros (or the equivalent `local_inner_macros`) and suddenly that means *you can't call your own macros anymore*, that seems like a critical issue.

**Comment by nikomatsakis** [comment]
@petrochenkov before, you wrote:

> With duplicate macro_exports prohibited this should be fixable if `$crate::foo!()` is defined in the same way as it's visible from other crates, i.e. referring to the corresponding unique macro_exported macro and not to one of the macros named foo and defined in the root module.

How hard do you think that would be?

Also, on a related note, if we *did* fix this, would that imply that `use crate::foo; foo!(...)` would *also* work? (where `foo` is a `#[macro_export]` macro defined somewhere in the crate?)

**Comment by petrochenkov** [comment]
>you can't call your own macros anymore, that seems like a critical issue.

Not really, it's [not common](https://github.com/rust-lang/rust/pull/51496#issuecomment-396327712) for a library to both define macros and use them outside of other macros, unless it's `libstd` or something like this.
The workaround is simple too, macros can be taken out in a separate crate, something that macro libraries already do casually due to procedural macros and derives requiring separate crates.

>How hard do you think that would be?

Not so simple.
The primary issue is breakage - every macro expansion in a crate can define a new `#[macro_export] macro_rules! { ... }` item and place it into the root, so we effectively have a permanently unexpanded macro call in the root until the very expansion end.
This means import resolution and expansion will get stuck in strictly more cases than now. Not sure how this may affect code in practice.
It would be interesting to experiment with this in the future, but it's in any way a blocker for stabilizing `use_extern_macros`.

>would that imply that `use crate::foo; foo!(...)` would also work? 

Yes.

**Comment by petrochenkov** [comment]
>How hard do you think that would be?

FWIW, for `#[macro_export(local_inner_macros)]` it's very easily fixable.
Since `local_inner_macros` is compiler magic, we can just avoid prepending `$crate::` if the macro is from our crate.

**Comment by joshtriplett** [comment]
@petrochenkov Good to know. But that would mean that `local_inner_macros` would be more than just "prefix every macro call with `$crate::`, and it'd be more than just a backwards-compatibility hack for crates supporting older rustc.

**Comment by joshtriplett** [comment]
@petrochenkov My concern is that crate maintainers will get requests to support `use_extern_macros`, leading them to switch to `$crate::` (or `local_inner_macros` for older compilers), and then they'll suddenly find they can't call their own macros (including from tests).

I guess I could live with `local_inner_macros` working for that, if we document that. It'd be a little unfortunate, though.

Does something make `$crate::` wildly harder to fix in that regard?

**Comment by petrochenkov** [comment]
@joshtriplett 
>Does something make `$crate::` wildly harder to fix in that regard?

Yes, see https://github.com/rust-lang/rust/pull/50911#issuecomment-401177651.
I'll try to come up with some prototype and run it through crater in the next 2-3 weeks, but if this remains the only blocker around August 3 (next release), we'd still be better stabilizing `use_extern_macros` and `proc_macro` without it.

**Comment by alexcrichton** [comment]
I'm catching up on related threads after coming back from vacation. I wanted to be sure to thank @dtolnay for [articulating the problem so well](https://github.com/rust-lang/rust/issues/35896#issuecomment-394553992) and @petrochenkov for [implementing a solution to this](https://github.com/rust-lang/rust/pull/51496), very nice work! Also thanks @petrochenkov for continuing to burn down macro-related issues in PRs like https://github.com/rust-lang/rust/pull/51762, https://github.com/rust-lang/rust/pull/51457, and https://github.com/rust-lang/rust/pull/51726. 

From what I understand one of the last known possible blockers is from [@dtolnay's analysis](https://github.com/rust-lang/rust/pull/51496#issuecomment-396315356) which is orthogonal to the new `local_inner_macros` attribute but rather [persists today on nightly with just `use_extern_macros`](https://play.rust-lang.org/?gist=3db3b5ef7d634365823d03ceb9d033c5&version=nightly&mode=debug). I agree with @dtolnay that this is likely uncommon enough to not come up enough in practice and it even [seems like there are ways to fix this after the fact](https://github.com/rust-lang/rust/pull/50911#issuecomment-401177651) (albeit it will be difficult as we need to move carefully to avoid regressions)

@petrochenkov you've also mentioned that tool attributes are blocking `use_extern_macro`, but can you clarify? My understanding of https://github.com/rust-lang/rust/issues/44690#issuecomment-398307401 and the discussion on that issue is that https://github.com/rust-lang/rust/issues/51277 is related but I'm not sure how that'd block `use_extern_macros` itself. I think that `#[rustfmt::anything]` is denied due to multi-segments right now and `#[rustfmt]` is also denied if `use_extern_macros` is enabled because there's no macro/attribute by that name, right?

**Comment by alexcrichton** [comment]
Oh sorry, misread!

@joshtriplett how do you feel about moving forward with stabilization with the known caveat of yours given the possible paths forward?

**Comment by joshtriplett** [comment]
As mentioned, if we made it so `local_inner_macros` still allowed you to call the macro from within your own crate, I could live with `$crate::` not working yet, though I'd like to see that fixed as well to avoid surprises.

We'd need to document that carefully.

**Comment by alexcrichton** [comment]
Ok! In that case ping r? @nikomatsakis 

I don't mind filling in documentation over the cycle!

**Comment by nrc** [comment]
My concern with stabilising now is that seems like a bit of a hack to get `use macro` in the edition release. Personally, I don't think that's even a great idea, and I don't think it is a certainty that the feature will make it. If we are going to wait anyway, are there nicer, long-term solutions to this problem that it's worth keeping this unstable for?

**Comment by alexcrichton** [comment]
@nrc hm I'm a bit confused, this has [gone through FCP](https://github.com/rust-lang/rust/issues/35896#issuecomment-378271865) for stabilization, so is the concerns you have those related to the new information brought to light between FCP and stabilization here? This is blocking procedural macros being stabilized as well

**Comment by nikomatsakis** [comment]
@alexcrichton I am not sure if that FCP still counts, since there was a significant problem (the macro solution), and in the process of discussing it, it seemed like feelings were shifting (@nrc's, in particular). Also, we've now changed the proposal to include the new feature, and we don't have a lot of experience with that.

(On the other hand, we don't have much time to waste here, clearly.)

**Comment by alexcrichton** [comment]
@nikomatsakis and/or @nrc would y'all be up for a call? I was still under the impression that we were set to have procedural macros be one of the major features of the 2018 edition, and it may be best to talk about concerns in real time?

**Comment by nikomatsakis** [comment]
I think the concerns have more to do `macro_rules!` style macros, and not procedural ones.

**Comment by cramertj** [comment]
@nikomatsakis Are you proposing that we stabilize `use` imports of procedural macros, but not macro_rules macros?

**Comment by nikomatsakis** [comment]
I am not yet proposing anything in particular, but I am saying that no matter what we decide to do with macro-rules macros, we can still go forward with procedural macros.

This is what I think our deadlines are:

- We should decide ASAP whether we want to have `use_extern_macro` for macro-rules macros as part of the edition.
  - I believe this is still mildly controversial.
- If not, we should immediately extract that to another feature gate and focus on procedural macros.
- If so, I don't think we have to stabilize **immediately** -- we will have various unstable features as part of the Edition Preview 2, and this can be among them I think?
  - In particular, we might want to gain more experience with the "local-inner-macro" thing, right?
  - Unless we want crates to adopt it?
- But I do think we should stabilize by the Release Candidate (1.29, as of now).

**Comment by alexcrichton** [comment]
In a short discussion with @nrc, @nikomatsakis, @joshtriplett, and I we concluded that the best way forward for now is to decouple the stabilization of procedural macros from macros 1.0 being available through the module system. 

I'm going to pursue a PR that stabilizes `use_extern_macros` *just* for procedural macros, not `macro_rules!` style macros.

**Comment by steveklabnik** [comment]
> the best way forward for now is to decouple the stabilization of procedural macros from macros 1.0 being available through the module system.

Does this mean anything for the macros 1.0 story? Or just purely practicality for this moment?

(`use` for `macro_rules` macros was one of the most desired features in the edition, so I'd like to know ASAP if the plan is to change it.)

**Comment by alexcrichton** [comment]
Good question! Technically there's no need to couple stabilization of things like `#[proc_macro]` and `#[proc_macro_attribute]` to the ability to `use` a `macro_rules!` macro. This next step is to perform the decoupling so we can stabilize `#[proc_macro]` and such. This does not mean, however, that `use` with `macro_rules!` won't happen!

Since the initial FCP the `local_inner_macros` option has been added for `macro_rules!` and stabilizing it now is a bit too quick on the stabilization timeline. Instead the plan is to shoot for 2018-08-02 for stabilizing procedural macros (this nightly cycle) and shoot for 2018-09-13 for `macro_rules!` and the module system. That means that the edition preview 2 should have procedural macros, and the release candidate should have `macro_rules!` and the module system.

It's hoped that an extra cycle is enough to test out and shake out bugs in `local_inner_macros`

**Comment by alexcrichton** [comment]
(although others should definitely correct me if I'm wrong!)

**Comment by cramertj** [comment]
I want to point out that this also means that many more uses of `extern crate` will not be removable since there won't be an alternative to `#[macro_use]` available.

**Comment by steveklabnik** [comment]
> I want to point out that this also means that many more uses of extern crate will not be removable since there won't be an alternative to #[macro_use] available.

Yes, this is a large part of why people are excited about the feature.

That said, you and Alex posted almost at the same time, and

>  Instead the plan is to shoot for 2018-08-02 for stabilizing procedural macros (this nightly cycle) and shoot for 2018-09-13 for macro_rules! and the module system. That means that the edition preview 2 should have procedural macros, and the release candidate should have macro_rules! and the module system.

This seems good to me!

**Comment by alexcrichton** [comment]
@cramertj sort of! We're still shooting for `macro_rules!`-and-the-module-system for the edition. I was personally hitching the `proc_macro` wagon trailer to this feature, but in our discussion it was discovered erroneously so. The most contentious part about stabilizing `use_extern_macros` is exclusively related to `macro_rules!`, so there's no need to block all of `proc_macro` stabilization on that. Instead we hope to give the recently implemented fix some more time to bake before it's stabilized

**Comment by cramertj** [comment]
@joshtriplett I'm successfully using the feature! :wink:

**Comment by alexcrichton** [comment]
@petrochenkov hm ok, I'll try to do this step by step in that case. I'm currently just trying to stabilize `proc_macro` the feature (and still require a feature gate to use them) and it's got tons of test failures. 

I don't think this is an impossible thing to decouple though? Naively I'd imagine that it's something like you should be able to import macros from `proc-macro` crates but no others, right?

**Comment by alexcrichton** [comment]
I've posted the first step in stabilizing procedural macros at https://github.com/rust-lang/rust/pull/52081

**Comment by petrochenkov** [comment]
>If we are not gating at all, but just "filter away" macro_rules from imports, then it would be not entirely backward compatible to not filter away them later due to possible new name conflicts in macro namespace. I think we can realistically do this in practice though if it's done soon enough.

I'll try to implement this variant of "partial ~stabilization~ enabling" for now and send a PR (+ later implement local `$crate::my_macro` for crater testing).
To clarify, this is a pretty bad temporary measure and it should be followed by full macro import ~stabilization~ enabling soon.

I'm also convinced that full `use_extern_macros` is ready for stabilization right now, since I was personally involved in reviewing/fixing all the related issues in the last couple of months, but if people want one more cycle for testing, then it's not a big deal, I guess, perhaps we will eliminate even more bugs.

**Comment by joshtriplett** [comment]
@petrochenkov To clarify something, I don't think anyone was arguing that we *have* to stabilize the two independently; the proposal at the lang team meeting was "we don't have to tie the two together if it's easier to handle them separately". However, if you find it easier to address both at once, I don't see any reason we couldn't do that. *Especially* if you see a path to addressing the local-macros case.

**Comment by petrochenkov** [comment]
PR for crate-local `$crate::macro_helper` - https://github.com/rust-lang/rust/pull/52234.

**Comment by alexcrichton** [comment]
@petrochenkov with all the recent PRs that have landed, did you have others in mind or is this ready to rebase/land? (we've also just started a new nightly cycle so this'd have plenty of time to bake)

**Comment by petrochenkov** [comment]
@alexcrichton 
>did you have others in mind or is this ready to rebase/land?

Yes, two so far.
- Support custom attributes when `use_extern_macros` is enabled (today-tomorrow).
- Route all attributes through resolution/expansion (https://github.com/rust-lang/rust/issues/52269) (most likely next weekend + maybe crater run).

After that we'll be able to land this PR, I think.

**Comment by alexcrichton** [comment]
Ok cool. I noticed that Serde broke with (what I believe) `use_extern_macros` enabled before your previous PRs landed. I think that the two more PRs you'd like to do won't affect most crates in general, right? If so, we could probably go ahead and schedule a crater run for a rebased version of this?

If you agree and you're busy, I don't mind doing the rebase here as well!

**Comment by petrochenkov** [comment]
I don't think it makes sense to run crater now, current version doesn't support custom attributes so there will be a lot of noise from this in results, even rustc cannot bootstrap without it.
So it makes sense to wait for the first PR at least.

(The second PR will [break stuff too](https://github.com/rust-lang/rust/issues/52269#issuecomment-407543177) and will need a separate crater run to figure out what we really need to unbreak, so there are many crater runs ahead :) )

**Comment by alexcrichton** [comment]
Sure thing, sounds good to me!

**Comment by nrc** [comment]
@petrochenkov do you have an ETA on this? Will it be possible to backport to beta? (It sounds like no, because there is a lot of complexity)

**Comment by alexcrichton** [comment]
@nrc oh we're definitely not gonna backport this, it's planned that we can land this for the 1.30 release though (today's nightly)

File: src/liballoc/boxed.rs
Comment (mati865) [inline src/liballoc/boxed.rs]: `#[rustfmt::skip]` can be used now, fixed by https://github.com/rust-lang/rust/pull/52841

**Comment by petrochenkov** [comment]
Rebase is done.

Also some status update.
- [Blocker] [ETA for the fix: 2 days] https://github.com/rust-lang/rust/pull/52234 caused a pretty bad regression due to causing indeterminate resolutions. It prevents rustc from bootstrapping without workarounds (see the examples with `log` and `lazy_static` in https://github.com/rust-lang/rust/pull/50911/commits/6c78d3b99f5056caf4f8289ee4a6559e6a915981) and other cases of breakage (https://github.com/rust-lang/rust/pull/53053#issuecomment-411564720, https://github.com/rust-lang/rust/issues/53144) were reported as well.
- [Blocker] [ETA for the fix: this weekend] As it was found in https://github.com/rust-lang/rust/pull/53044 `use_extern_macros` now also gates arbitrary tokens in non-macro attributes. This needs a separate feature gate + attribute grammar needs to be actually enforced by the parser rather than by random pieces of code here and there with few pieces missing.
- [Blocker] [ETA for the fix: this weekend] Check whether proc macros can be used in inner attributes, feature gate or prohibit if necessary. Inner attributes also live in a different module/block then their outer neighbors and so should be resolved from other scope, but this isn't currently done AFAIK.
- [Blocker] [ETA for the fix: next week] Prohibit macro attributes in locations that don't support macro expansion (e.g `fn f<#[attr] T>() {}`).
- [Blocker] [ETA for the fix: next week] Pass possibly-builtin attributes through name resolution. I started working on this, but going to postpone a bit to fix the regression from https://github.com/rust-lang/rust/pull/52234 first.
- [Optional] [ETA: after landing this PR until the upcoming beta] Implement and crater a principled left-to-right scheme for attribute expansion and derive helper attribute resolution (https://github.com/rust-lang/rust/issues/52226).

All the items are relatively small in terms of implementation, so we should be able to do the final crater run around the next weekend and then hopefully land this PR.

**Comment by alexcrichton** [comment]
Thanks for the updates here @petrochenkov! I'm currently traveling right now but should have more time next week or the week after, and I'd gladly take the smaller items that don't require much resolve knowledge (like gating and such), so don't feel pressured to take those on!

File: src/librustc_resolve/macros.rs
Comment (petrochenkov) [inline src/librustc_resolve/macros.rs]: FIXME: `lexical_macro_resolutions` is now unused

**Comment by alexcrichton** [comment]
@bors: r+

Sounds like a solid plan to me, thanks @petrochenkov!

**Comment by alexcrichton** [comment]
🎊 

Thanks so much for helping this over the finish line @petrochenkov!

**Comment by softprops** [comment]
The [1.30.0 release notes](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1300-2018-10-25) for `#[macro_export(local_inner_macros)]` link here.

Where can I find more documentation on this feature?

So far I've tried local_inner_macros in two  cases where a crate's macros depended on anothers and it didn't work in either case.

https://github.com/ilianaw/rust-crowbar/pull/46#issuecomment-433422255

https://github.com/dtolnay/mashup/issues/16

**Comment by petrochenkov** [comment]
@softprops 
https://github.com/rust-lang-nursery/edition-guide/pull/117

It will be available at https://rust-lang-nursery.github.io/edition-guide/rust-2018/macros/macro-changes.html when merged.

**Comment by softprops** [comment]
@petrochenkov thanks! I'll take a look and report back if it looks like what I'm trying to do, re-export transitive dependency crates macros` is actually what this new feature supports.

**Comment by softprops** [comment]
@petrochenkov l need to do a bit more testing but I worked it out, at least enough to get a fork of rust crowbar to compile. I still need to test a crate that uses that fork before I can call this 👍 'd

It may not be clear to new rust users that for foreign macros, you'll also need namespace the macro's path. In my case I just needed to prefix crowbars use of cpython's macros with `cpython::` in the docs you linked to and users will find you only use local macros in examples  using `$crate::`. Assuming macros authors are already going to be a bit more advanced than the beginner rust users they may be able to discover this for themselves.  I'd recommend adding at least one foreign macro example for the sake of beginner/advanced user inclusivity. Albeit this would be _much_ more painful for the example of supporting old rust versions because if forces users to copy and paste the entire macro srcs from other crates. The `format_args` is probably the simplest case for a macro. A more complex examples will be much more awkward. For now, asking users to use the latest stable works well enough!

I reall

**Comment by petrochenkov** [comment]
@softprops 
I suspect that the `lambda` macro is recursive and calls itself by relative name (`lambda!(...)`).

This way the first iteration `crowbar::lambda!(...)` is resolved successfully, but the next iterations cannot be resolved without bringing `lambda` in scope at use-site or using `local_inner_macros` at def-site.

**Comment by softprops** [comment]
@petrochenkov 100%. that was it. thanks!

**Pull request merged**
